rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,275 @@
+# RFC 5843 - 2021-01-08 - Encoding/Decoding for Vector to Vector Communication
+
+There has been an ongoing discussion of what format Vector-to-Vector communications should take. For some examples of issues related to the subject check out #5809, #5124, or #5341. This RFC is intended to propose a format for standardization of our vector-to-vector communications.
+
+## Scope
+
+- This RFC should cover the encoding format we intend to leverage
+- It should also cover any alterations that might need to be made to the expected data format
+
+## Motivation
+
+The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing an undocumented format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.
```

Comment:
It's a bit odd. I think it's  _technically_ a "public" API. It's what we use to make the vector source and vector sink talk to each other and I'm assuming that users could push to or pull from this source and sink, but I don't know _why_ someone would want to. Though I'd actually probably lean on someone with a more thorough internals knowledge  than myself to fill in the gaps on that (slash correct me if I'm way off base).

Reply:
I ask because if we document the API as being internal-only then we're beholden only to ourselves for maintaining compatibility. That gives us more freedom to experiment with an approach and back off it if we need to, is why I ask.

Reply:
Absolutely. I'm not certain if @lukesteensen has any feelings on this. I'd hate to limit our users from something that could be useful for them but I have a hard time seeing this as anything other than an internal communication API and as such I think it makes sense to document this as internal only.

Reply:
The intent is definitely for this to be internal-only. We don't want to introduce yet another log/metric data format into the world, but we do want something that maps pretty directly to our internal data representation and allows us the flexibility to evolve that representation as needed.